File: app/src/main/java/com/example/taskapp/ui/TasksViewModel.java
package com.example.taskapp.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.example.taskapp.data.ApiResponse;
import com.example.taskapp.data.Task;
import com.example.taskapp.data.TaskRepository;
import java.util.List;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class TasksViewModel extends ViewModel {
    private TaskRepository repository;
    private MutableLiveData<Boolean> loading = new MutableLiveData<>();
    private MutableLiveData<String> error = new MutableLiveData<>();
    private MutableLiveData<List<Task>> tasks = new MutableLiveData<>();

    public TasksViewModel(TaskRepository repository) {
        this.repository = repository;
    }

    public LiveData<Boolean> getLoading() { return loading; }
    public LiveData<String> getError() { return error; }
    public LiveData<List<Task>> getTasks() { return tasks; }

    public void fetchTasks() {
        Call<List<Task>> call = repository.getTasks();
        if (call == null) return;
        loading.setValue(true);
        call.enqueue(new Callback<List<Task>>() {
            @Override
            public void onResponse(Call<List<Task>> call, Response<List<Task>> response) {
                if (response.isSuccessful()) {
                    tasks.setValue(response.body());
                } else {
                    error.setValue("Failed to fetch tasks: " + response.code());
                }
                loading.setValue(false);
            }

            @Override
            public void onFailure(Call<List<Task>> call, Throwable t) {
                error.setValue(t.getMessage());
                loading.setValue(false);
            }
        });
    }

    public void createTask(String title) {
        Call<Task> call = repository.createTask(title);
        if (call == null) return;
        loading.setValue(true);
        call.enqueue(new Callback<Task>() {
            @Override
            public void onResponse(Call<Task> call, Response<Task> response) {
                if (response.isSuccessful()) {
                    fetchTasks();
                } else {
                    error.setValue("Failed to create task: " + response.code());
                }
                loading.setValue(false);
            }

            @Override
            public void onFailure(Call<Task> call, Throwable t) {
                error.setValue(t.getMessage());
                loading.setValue(false);
            }
        });
    }

    public void toggleTask(Task task) {
        Call<Task> call = repository.updateTask(task.getId(), !task.isCompleted());
        if (call == null) return;
        call.enqueue(new Callback<Task>() {
            @Override
            public void onResponse(Call<Task> call, Response<Task> response) {
                if (response.isSuccessful()) {
                    fetchTasks();
                } else {
                    error.setValue("Failed to update task: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<Task> call, Throwable t) {
                error.setValue(t.getMessage());
            }
        });
    }

    public void updateTaskTitle(Task task, String newTitle) {
        Call<Task> call = repository.updateTaskTitle(task.getId(), newTitle);
        if (call == null) return;
        loading.setValue(true);
        call.enqueue(new Callback<Task>() {
            @Override
            public void onResponse(Call<Task> call, Response<Task> response) {
                if (response.isSuccessful()) {
                    fetchTasks();
                } else {
                    error.setValue("Failed to update task title: " + response.code());
                }
                loading.setValue(false);
            }

            @Override
            public void onFailure(Call<Task> call, Throwable t) {
                error.setValue(t.getMessage());
                loading.setValue(false);
            }
        });
    }

    public void deleteTask(Task task) {
        Call<ApiResponse> call = repository.deleteTask(task.getId());
        if (call == null) return;
        call.enqueue(new Callback<ApiResponse>() {
            @Override
            public void onResponse(Call<ApiResponse> call, Response<ApiResponse> response) {
                if (response.isSuccessful()) {
                    fetchTasks();
                } else {
                    error.setValue("Failed to delete task: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<ApiResponse> call, Throwable t) {
                error.setValue(t.getMessage());
            }
        });
    }

    public void clearError() {
        error.setValue(null);
    }
}

File: app/src/main/java/com/example/taskapp/ui/TasksViewModel.kt
package com.example.taskapp.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.taskapp.data.Task
import com.example.taskapp.data.TaskRepository
import kotlinx.coroutines.launch

class TasksViewModel(private val repository: TaskRepository) : ViewModel() {

    private val _loading = MutableLiveData<Boolean>()
    val loading: LiveData<Boolean> get() = _loading

    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> get() = _error

    private val _tasks = MutableLiveData<List<Task>>()
    val tasks: LiveData<List<Task>> get() = _tasks

    fun fetchTasks() {
        _loading.value = true
        viewModelScope.launch {
            try {
                val response = repository.getTasks()
                if (response.isSuccessful) {
                    _tasks.postValue(response.body() ?: emptyList())
                } else {
                    _error.postValue("Failed to fetch tasks: ${response.code()}")
                }
            } catch (e: Exception) {
                _error.postValue(e.message ?: "An error occurred")
            } finally {
                _loading.postValue(false)
            }
        }
    }

    fun createTask(title: String) {
        _loading.value = true
        viewModelScope.launch {
            try {
                val response = repository.createTask(title)
                if (response.isSuccessful) {
                    fetchTasks()
                } else {
                    _error.postValue("Failed to create task: ${response.code()}")
                }
            } catch (e: Exception) {
                _error.postValue(e.message ?: "An error occurred")
            } finally {
                _loading.postValue(false)
            }
        }
    }

    fun toggleTask(task: Task) {
        viewModelScope.launch {
            try {
                val response = repository.updateTask(task.id, !task.completed)
                if (response.isSuccessful) {
                    fetchTasks()
                } else {
                    _error.postValue("Failed to update task: ${response.code()}")
                }
            } catch (e: Exception) {
                _error.postValue(e.message ?: "An error occurred")
            }
        }
    }

    fun updateTaskTitle(task: Task, newTitle: String) {
        _loading.value = true
        viewModelScope.launch {
            try {
                val response = repository.updateTaskTitle(task.id, newTitle)
                if (response.isSuccessful) {
                    fetchTasks()
                } else {
                    _error.postValue("Failed to update task title: ${response.code()}")
                }
            } catch (e: Exception) {
                _error.postValue(e.message ?: "An error occurred")
            } finally {
                _loading.postValue(false)
            }
        }
    }

    fun deleteTask(task: Task) {
        viewModelScope.launch {
            try {
                val response = repository.deleteTask(task.id)
                if (response.isSuccessful) {
                    fetchTasks()
                } else {
                    _error.postValue("Failed to delete task: ${response.code()}")
                }
            } catch (e: Exception) {
                _error.postValue(e.message ?: "An error occurred")
            }
        }
    }

    fun clearError() {
        _error.value = null
    }
}
